dataflow.py

```diff
@@ -1,81 +1,98 @@
 import json
-
+from collections import OrderedDict
+from dataclasses import dataclass, field
+from datetime import timedelta
+from typing import List, Optional
+
+# pip install websockets
+import websockets
+from bytewax import operators as op
+from bytewax.connectors.stdio import StdOutSink
 from bytewax.dataflow import Dataflow
-from bytewax.inputs import DynamicInput, StatelessSource
-from bytewax.connectors.stdio import StdOutput
-from websocket import create_connection  # pip install websocket-client
-
-
-class CoinbaseSource(StatelessSource):
-    def __init__(self, product_ids):
-        self.product_ids = product_ids
-
-        self.ws = create_connection("wss://ws-feed.pro.coinbase.com")
-        self.ws.send(
-            json.dumps(
-                {
-                    "type": "subscribe",
-                    "product_ids": product_ids,
-                    "channels": ["level2_batch"],
-                }
-            )
+from bytewax.inputs import FixedPartitionedSource, StatefulSourcePartition, batch_async
+
+
+async def _ws_agen(product_id):
+    url = "wss://ws-feed.exchange.coinbase.com"
+    async with websockets.connect(url) as websocket:
+        msg = json.dumps(
+            {
+                "type": "subscribe",
+                "product_ids": [product_id],
+                "channels": ["level2_batch"],
+            }
         )
+        await websocket.send(msg)
         # The first msg is just a confirmation that we have subscribed.
-        print(self.ws.recv())
+        await websocket.recv()
 
-    def next(self):
-        return self.ws.recv()
+        while True:
+            msg = await websocket.recv()
+            yield (product_id, json.loads(msg))
 
 
-class CoinbaseInput(DynamicInput):
-    PRODUCT_IDS = ["BTC-USD", "ETH-USD", "SOL-USD"]
+class CoinbasePartition(StatefulSourcePartition):
+    def __init__(self, product_id):
+        agen = _ws_agen(product_id)
+        self._batcher = batch_async(agen, timedelta(seconds=0.5), 100)
 
-    def build(self, worker_index, worker_count):
-        prods_per_worker = int(len(self.PRODUCT_IDS) / worker_count)
-        product_ids = self.PRODUCT_IDS[
-            int(worker_index * prods_per_worker) : int(
-                worker_index * prods_per_worker + prods_per_worker
-            )
-        ]
-        return CoinbaseSource(product_ids)
+    def next_batch(self):
+        return next(self._batcher)
 
+    def snapshot(self):
+        return None
 
-flow = Dataflow()
-flow.input("input", CoinbaseInput())
 
```

Comment:
This is just to let you know why the code in the bytewax repo examples is styled differently than the guides. The guides' side-by-side comparison presents a different experience viewing the code than a plain file. As you scroll through the sections it jumps all over the place in the code to the right lines. For this reason, I ended up writing the code in more of a literate or notebook style.

`function`
`operation`

So that when the section was clicked on in the guide it could talk about both the input class and the operator or the map function and the map operator without splitting them up or jumping around.

I am not sure what is the right style to use, but that's where I ended up.

Reply:
Understand